File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/profile/Windows.java
package gov.nasa.jpl.aerie.constraints.profile;

import gov.nasa.jpl.aerie.constraints.time.Interval;
import gov.nasa.jpl.aerie.constraints.time.IntervalContainer;
import gov.nasa.jpl.aerie.constraints.time.Segment;
import gov.nasa.jpl.aerie.constraints.time.Spans;
import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;

import java.util.Collections;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;

public interface Windows extends Profile<Boolean>, IntervalContainer<Windows> {
  static Windows from(final boolean constant) {
    return Profile.from(constant)::stream;
  }

  default Windows not() {
    return mapValues($ -> !$)::stream;
  }

  default Windows and(final Profile<Boolean> other) {
    return Profile.map2Values(
        this, other,
        BinaryOperation.fromCases(
            l -> l ? Optional.empty() : Optional.of(false),
            r -> r ? Optional.empty() : Optional.of(false),
            (l, r) -> Optional.of(l && r)
        )
    )::stream;
  }

  default Windows or(final Profile<Boolean> other) {
    return Profile.map2Values(
        this, other,
        BinaryOperation.fromCases(
            l -> l ? Optional.of(true) : Optional.empty(),
            r -> r ? Optional.of(true) : Optional.empty(),
            (l, r) -> Optional.of(l || r)
        )
    )::stream;
  }

  default Windows add(final Profile<Boolean> other) {
    return Profile.map2Values(
        this, other,
        BinaryOperation.combineDefaultIdentity((l, r) -> l || r)
    )::stream;
  }

  default Windows trueSubsetContainedIn(final Interval interval) {
    return bounds -> this.stream(bounds).map(
      $ -> {
        if (interval.contains($.interval())) {
          return $;
        } else {
          return $.mapValue(v -> false);
        }
      }
    );
  }

  default Windows removeTrueSegment(final int index) {
    return bounds -> {
      final Stream<Segment<Boolean>> reversedStream;
      final int reversedIndex;
      if (index >= 0) {
        reversedStream = this.stream(bounds);
        reversedIndex = index;
      } else {
        final var list = this.stream(bounds).toList();
        Collections.reverse(list);
        reversedStream = list.stream();
        reversedIndex = -index - 1;
      }
      final AtomicInteger segmentCounter = new AtomicInteger(0);
      final var intermediateStream = reversedStream.map(
          $ -> {
            if ($.value() && segmentCounter.getAndIncrement() == index) {
              return $.mapValue(v -> false);
            } else {
              return $;
            }
          }
      );
      if (index >= 0) {
        return intermediateStream;
      } else {
        final var list = intermediateStream.toList();
        Collections.reverse(list);
        return list.stream();
      }
    };
  }

  default Windows keepTrueSegment(final int index) {
    return bounds -> {
      final Stream<Segment<Boolean>> reversedStream;
      final int reversedIndex;
      if (index >= 0) {
        reversedStream = this.stream(bounds);
        reversedIndex = index;
      } else {
        final var list = this.stream(bounds).toList();
        Collections.reverse(list);
        reversedStream = list.stream();
        reversedIndex = -index - 1;
      }
      final AtomicInteger segmentCounter = new AtomicInteger(0);
      final var intermediateStream = reversedStream.map(
          $ -> {
            if ($.value() && segmentCounter.getAndIncrement() != index) {
              return $.mapValue(v -> false);
            } else {
              return $;
            }
          }
      );
      if (index >= 0) {
        return intermediateStream;
      } else {
        final var list = intermediateStream.toList();
        Collections.reverse(list);
        return list.stream();
      }
    };
  }

  default boolean includes(final Profile<Boolean> other, final Interval bounds) {
    final var extraSegments = Profile.map2Values(
        this, other,
        BinaryOperation.fromCases(
            $ -> Optional.of(false),
            Optional::of,
            (l, r) -> Optional.of(r && !l)
        )
    );

    // anywhere where the above has true means other was not included in this.
    return extraSegments.stream(bounds).noneMatch(Segment::value);
  }

  default Windows filterByDuration(final Duration min, final Duration max) {
    return bounds -> this.stream(bounds).map(
      $ -> {
        if ($.value()) {
          final var duration = $.interval().duration();
          if (duration.shorterThan(min) || duration.longerThan(max)) {
            return Segment.of($.interval(), false);
          } else {
            return $;
          }
        } else {
          return $;
        }
      }
    );
  }

  default Windows shiftEdges(final Duration shiftRising, final Duration shiftFalling) {
    return bounds -> {
      final var newBounds = Interval.between(
          Duration.min(bounds.start.minus(shiftRising), bounds.start.minus(shiftFalling)),
          bounds.startInclusivity,
          Duration.max(bounds.end.minus(shiftRising), bounds.end.minus(shiftFalling)),
          bounds.endInclusivity
      );
      return this.stream(newBounds).map(
          $ -> {
            if ($.value()) {
              return $.mapInterval(
                  interval -> Interval.between(
                      interval.start.saturatingPlus(shiftRising),
                      interval.startInclusivity,
                      interval.end.saturatingPlus(shiftFalling),
                      interval.endInclusivity
                  )
              );
            } else {
              return $.mapInterval(
                  interval -> Interval.between(
                      interval.start.saturatingPlus(shiftFalling),
                      interval.startInclusivity,
                      interval.end.saturatingPlus(shiftRising),
                      interval.endInclusivity
                  )
              );
            }
          }
      ).filter($ -> !$.interval().isEmpty());
    };
  }

  @Override
  default Windows starts() {
    return specificEdges(false, true);
  }

  @Override
  default Windows ends() {
    return specificEdges(true, false);
  }

  default Spans intoSpans(final Interval bounds) {
    return new Spans(
        this.stream(bounds)
            .filter(Segment::value)
            .map(Segment::interval)
    );
  }
}

File: merlin-server/constraints-dsl-compiler/src/libs/windows.ts
import {Profile} from "./profile";
import {Segment} from "./segment";
import database from "./database";
import {BinaryOperation} from "./binary-operation";
import {Interval} from "./interval";
import type {Timeline} from "./timeline";
import {ProfileType} from "./profile-type";

export class Windows extends Profile<boolean> {
  constructor(segments: Timeline<Segment<boolean>>) {
    super(segments, ProfileType.Windows);
  }

  public static empty(): Windows {
    return new Windows(_ => []);
  }

  public static override Value(value: boolean, interval?: Interval): Windows {
    return new Windows(bounds => [new Segment(
        value,
        interval === undefined ? bounds : Interval.intersect(bounds, interval)
    )]);
  }

  public static override Resource(name: string): Windows {
    return new Windows(database.getResource(name));
  }

  public not(): Windows {
   return this.mapValues($ => !$);
  }

  public and(other: Windows): Windows {
    return this.map2Values(other, BinaryOperation.cases(
        l => l ? undefined : false,
        r => r ? undefined : false,
        (l, r) => l && r
    ), ProfileType.Windows);
  }

  public or(other: Windows): Windows {
    return this.map2Values(other, BinaryOperation.cases(
        l => l ? true : undefined,
        r => r ? true : undefined,
        (l, r) => l || r
    ), ProfileType.Windows);
  }

  public add(other: Windows): Windows {
    return this.map2Values(other, BinaryOperation.combineOrIdentity(
        (l, r) => l || r
    ), ProfileType.Windows);
  }

  public falsifyByDuration(min?: Temporal.Duration, max?: Temporal.Duration): Windows {
    return this.mapValues((v, i) => {
      if (v) {
        const duration = i.duration();
        return !((min !== undefined && Temporal.Duration.compare(min, duration) > 0) || (max !== undefined && Temporal.Duration.compare(max, duration) < 0));
      } else {
        return false;
      }
    });
  }

  public shorterThan(max: Temporal.Duration): Windows {
    return this.falsifyByDuration(undefined, max);
  }

  public longerThan(min: Temporal.Duration): Windows {
    return this.falsifyByDuration(min, undefined);
  }

  public shiftBy(shiftRising: Temporal.Duration, shiftFalling?: Temporal.Duration): Windows {
    if (shiftFalling === undefined) shiftFalling = shiftRising;
    const boundsMap = (bounds: Interval) => {
      let start: Temporal.Duration;
      let end: Temporal.Duration;
      if (Temporal.Duration.compare(shiftRising, shiftFalling!) === 1) {
        start = bounds.start.subtract(shiftRising);
        end = bounds.end.subtract(shiftFalling!);
      } else {
        start = bounds.start.subtract(shiftFalling!);
        end = bounds.end.subtract(shiftRising);
      }
      return Interval.between(
          start,
          end,
          bounds.startInclusivity,
          bounds.endInclusivity
      );
    }
    return this.unsafe.mapIntervals((v, i) => {
      if (v) {
        return Interval.between(
          i.start.add(shiftRising),
          i.end.add(shiftFalling!),
          i.startInclusivity,
          i.endInclusivity
        );
      } else {
        return Interval.between(
          i.start.add(shiftFalling!),
          i.end.add(shiftRising),
          i.startInclusivity,
          i.endInclusivity
        );
      }
    }, boundsMap);
  }

  public starts = () => this.transitions(false, true);
  public ends = () => this.transitions(true, false);
}
